design/proposal-crossplane-v2.md

```diff
@@ -0,0 +1,948 @@
+spec:
+  scope: Namespaced
+  group: platform.example.org
+  names:
+    kind: Application
+    plural: applications
+  versions:
+  - name: v1alpha1
+  # Removed for brevity
+```
+
+In Crossplane v2 a namespaced XR would only compose resources in its own
+namespace. It  wouldn’t support composing resources in other namespaces, or at
+the cluster scope. [This is a Kubernetes limitation](https://github.com/kubernetes/community/blob/95af16e046ae65d7c9d6ba68eb1389bf8462301f/contributors/devel/sig-architecture/api-conventions.md#object-references).
+A namespaced resource can’t own a resource at the cluster scope, or in another
+namespace:
+
+_Built-in types and ownerReferences do not support cross namespaces references._
+
+A cluster scoped XR would be able to compose resources at the cluster scope, or
+in any namespace(s).
+
+With these changes, XRs would act more like idiomatic Kubernetes resources. I
+expect that most XRs would become namespaced over time, and would represent
+applications or infrastructure owned by application developers.
+
+Cluster scoped XRs would be an edge case, just like cluster scoped resources are
+an edge case in Kubernetes. I expect cluster scoped XRs would be used for APIs
+that configure the control plane, for example by composing RBAC
+ClusterRoleBindings, or CustomResourceDefinitions.
+
+XRDs and Compositions would remain cluster scoped, as would the Configuration
+package type that you use to install them.
+
+#### Compose Any Kubernetes Resource
+
+I propose that Crossplane v2 allow XRs to compose any type of Kubernetes
+resource - not only Crossplane resources like MRs and XRs.
+
+This is more of a change of position than a technical change. There’s no
+technical reason you can’t compose arbitrary Kubernetes resources in Crossplane
+today except that those resources must be cluster scoped, and most Kubernetes
+resources aren’t. By introducing namespaced XRs that compose namespaced
+resources we’d implicitly support composing any Kubernetes resource.
+
+Composing arbitrary resources is a tradeoff.
+
+When the resources Crossplane composes become less consistent, XRs become less
+predictable. All Crossplane MRs have consistent schemas and behaviors:
+
+* They all support consistent external identity (the
+  [`crossplane.io/external-name`](http://crossplane.io/external-name)
+  annotation)
+* They all support consistent publishing of secret connection details
+* They all support consistent, predictable status conditions
+* They all support management policies to enable observe-only mode,
+  orphan-on-delete, etc
+
+MRs are all built using the same reconciler library. This means all MRs have the
+same retry, polling, and rate limiting semantics
+
+Controlling the entire stack allows Crossplane to serve its user base better.
+Consider management policies, which allows any Crossplane MR to be put into an
+observe-only mode (among other things). It’s controlling the MR layer that
```

Comment:
How would crossplane manage observe-only resources for in-cluster MRs in this manner and how would that differ from the behaviour of say `function-extra-resources` Would an in-cluster O-O MR be able to observe across namespace or would you imagine this being entirely restricted to its own namespace